stresstests/blobfusestress.cpp

```diff
@@ -0,0 +1,619 @@
+// Basically, we start a bunch (constant number) of std::thread threads, and store them in m_threads.
+// We also store a deque (double-ended queue) of work to do, where a unit of work is some arbitrary std::function<void()>.  Work is performed in roughly FIFO order.
+// Each thread runs one work item at a time.  When it finished it tries to get another unit of work, or goes to sleep.
+// Work is added to the queue via the add_task method.
+// Note that there is only one queue of work, and all operations to it (pushing new work and popping work off to perform it) require mutex operations.  
+// This may be a significant amount of overhead if the size of the work is very small and the number of work items is very large.  So for example, if the work is to copy
+// a million very small files, it will be very bad to push the copy of each file to the threadpool as a separate item.  Much better to chunk it externally.
+class thread_pool {
+public:
+
+    // Loop that each thread runs, to pop work off the queue and execute it.
+    void run_thread(int thread_id)
+    {
+        // m_finished is set during threadpool destruction, allowing the threads to all exit gracefully.
+        while (!m_finished)
+        {
+            std::function<void()> work;
+            {
+                // Lock the mutex for the duration of the queue operations, but not performing the actual work.
+                std::unique_lock<std::mutex> lk(m_mutex);
+                m_wait_count++;
+                // Release the mutex and perhaps sleep until woken up and there is work to be done (or tearing down the threadpool.)
+                m_cv.wait(lk, [this] {return (!m_task_list.empty()) || m_finished;});
+                m_wait_count--;
+                if (!m_finished)
+                {
+                    // Pop a unit of work from the queue; we still hold the mutex at this point.
+                    work = m_task_list.front();
+                    m_task_list.pop_front();
+                }
+            }
+
+            // Actually do the work.
+            if (!m_finished)
+            {
+                work();
+            }
+        }
+    }
+
+    // Spins up the desired number of threads, each of which will immediately wait on the condition variable.
+    thread_pool(int size)
+     : m_threads(), m_task_list(), m_mutex(), m_cv(), m_wait_count(0), m_finished(0)
+    {
+        for (int i = 0; i < size; i++)
+        {
+            m_threads.push_back(std::thread(&thread_pool::run_thread, this, i));
+        }
+    }
+
+    // Locks the work queue, adds a task to it, and then wakes up a worker thread to do the work.
+    // If no worker threads are waiting, notify_one is a no-op.
+    void add_task(std::function<void()> task)
+    {
+        std::lock_guard<std::mutex> lk(m_mutex);
+        {
+            m_task_list.push_back(task);
+        }
+        m_cv.notify_one();
```

Comment:
You may want notify all here? Say there are three threads and five tasks. Tasks 1, 2, and 3 are picked up immediately and take a while. Task 4 is placed in the queue while all three threads are busy. notify_one is called and is a no_op. Thread 1 finishes task 1. Task 5 is placed in the queue. notify_one is called and thread 1 takes task 4. All tasks have been added, but there are no more calls to notify_one, so task 5 lingers. Drain seems to assume that all remaining work has been picked up out of the queue already. It never wakes anyone up to clean up any lingering items.